File: jar/src/main/java/jline/solvers/fluid/analyzers/MFQAnalyzer.java
package jline.solvers.fluid.analyzers;

import jline.GlobalConstants;
import jline.lang.NetworkStruct;
import jline.lang.constant.NodeType;
import jline.lib.butools.queues.FluFluQueue;
import jline.lib.butools.queues.FluFluResult;
import jline.solvers.SolverOptions;
import jline.solvers.SolverResult;
import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

/**
 * MFQ (Markovian Fluid Queue) analyzer for single-queue open systems.
 */
public class MFQAnalyzer implements FluidAnalyzer {
    private Matrix xvecIt;

    @Override
    public void analyze(NetworkStruct sn, SolverOptions options, SolverResult result) {
        int M = sn.nstations;
        int K = sn.nclasses;
        result.QN = Matrix.zeros(M, K);
        result.UN = Matrix.zeros(M, K);
        result.RN = Matrix.zeros(M, K);
        result.TN = Matrix.zeros(M, K);
        result.WN = new Matrix(0, 0);
        result.AN = new Matrix(0, 0);

        TopologyInfo topologyInfo = validateTopology(sn);
        if (!topologyInfo.isValid) {
            throw new RuntimeException("MFQ requires single-queue topology: " + topologyInfo.errorMsg);
        }
        int sourceStation = topologyInfo.sourceStation;
        int queueStation = topologyInfo.queueStation;

        for (int k = 0; k < K; k++) {
            if (sn.njobs.get(0, k) == Double.POSITIVE_INFINITY) {
                ClassResult classResult = analyzeOpenClass(sn, k, sourceStation, queueStation, options);
                result.QN.set(queueStation, k, classResult.meanQueueLength);
                result.RN.set(queueStation, k, classResult.meanResponseTime);
                result.TN.set(queueStation, k, classResult.throughput);
                result.UN.set(queueStation, k, classResult.utilization);
                result.TN.set(sourceStation, k, classResult.throughput);
            }
        }
        xvecIt = Matrix.zeros(1, 1);
    }

    @Override
    public Matrix getXVecIt() { return xvecIt; }

    /**
     * The reason MFQ cannot run on this model, or {@code null} when it can.
     *
     * MFQ IS A SINGLE-QUEUE METHOD AND FALLS BACK: `solver_fluid_analyzer.m`
     * warns "MFQ not applicable: ... Falling back to matrix method" and
     * re-enters `solver_fluid_matrix`. The decision has to be taken by the
     * CALLER, before the mfq branch of runAnalyzer diverts around the state-space
     * preparation the matrix method reads (options.init_sol above all), which is
     * why this is exposed rather than handled inside {@link #analyze}.
     *
     * @param sn the model
     * @return the topology error, or null when MFQ applies
     */
    public static String mfqNotApplicableReason(NetworkStruct sn) {
        TopologyInfo info = new MFQAnalyzer().validateTopology(sn);
        return info.isValid ? null : info.errorMsg;
    }

    private TopologyInfo validateTopology(NetworkStruct sn) {
        TopologyInfo info = new TopologyInfo();
        boolean hasOpenClass = false;
        for (int k = 0; k < sn.nclasses; k++) {
            if (sn.njobs.get(0, k) == Double.POSITIVE_INFINITY) { hasOpenClass = true; break; }
        }
        if (!hasOpenClass) { info.errorMsg = "Not an open model - all classes are closed"; return info; }

        int sourceNode = -1, queueNode = -1, sinkNode = -1;
        int sourceCount = 0, queueCount = 0, sinkCount = 0;
        for (int i = 0; i < sn.nnodes; i++) {
            NodeType nt = sn.nodetype.get(i);
            if (nt == NodeType.Source) { sourceNode = i; sourceCount++; }
            else if (nt == NodeType.Queue) { queueNode = i; queueCount++; }
            else if (nt == NodeType.Sink) { sinkNode = i; sinkCount++; }
        }

        if (sourceCount == 0) { info.errorMsg = "No source node found"; return info; }
        if (sourceCount > 1) { info.errorMsg = "Multiple source nodes found (" + sourceCount + ")"; return info; }
        if (sinkCount == 0) { info.errorMsg = "No sink node found"; return info; }
        if (sinkCount > 1) { info.errorMsg = "Multiple sink nodes found (" + sinkCount + ")"; return info; }
        if (queueCount == 0) { info.errorMsg = "No queue node found"; return info; }
        if (queueCount > 1) { info.errorMsg = "Multiple queue nodes found (" + queueCount + ") - MFQ supports single queue only"; return info; }

        int sourceStation = (int) sn.nodeToStation.get(sourceNode);
        int queueStation = (int) sn.nodeToStation.get(queueNode);
        double c = sn.nservers.get(queueStation, 0);
        if (c > 1 && c < Double.POSITIVE_INFINITY) {
            info.errorMsg = "Multi-server queue (c=" + c + ") not supported - MFQ requires c=1 or c=Inf";
            return info;
        }
        info.isValid = true;
        info.sourceNode = sourceNode;
        info.queueNode = queueNode;
        info.sinkNode = sinkNode;
        info.sourceStation = sourceStation;
        info.queueStation = queueStation;
        return info;
    }

    private ClassResult analyzeOpenClass(NetworkStruct sn, int classIdx, int sourceStation, int queueStation, SolverOptions options) {
        double lambda = sn.rates.get(sourceStation, classIdx);
        if (Double.isNaN(lambda) || lambda <= 0) {
            throw new RuntimeException("No valid arrival rate for class " + classIdx + " at source");
        }
        double mu = sn.rates.get(queueStation, classIdx);
        if (Double.isNaN(mu) || mu <= 0) {
            throw new RuntimeException("No valid service rate for class " + classIdx + " at queue");
        }
        double rho = lambda / mu;
        if (rho >= 1.0) {
            return new ClassResult(Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY, lambda, 1.0);
        }
        ProcessParams arrivalParams = extractProcess(sn, sourceStation, classIdx, lambda);
        ProcessParams serviceParams = extractProcess(sn, queueStation, classIdx, mu);
        boolean isSimpleExponential = arrivalParams.isSimple && serviceParams.isSimple;

        double[] flMoms;
        double[] stMoms;
        if (isSimpleExponential) {
            double meanL = rho / (1 - rho);
            double varL = rho / ((1 - rho) * (1 - rho));
            double meanW = 1.0 / (mu - lambda);
            double varW = 1.0 / ((mu - lambda) * (mu - lambda));
            flMoms = new double[]{meanL, meanL * meanL + varL};
            stMoms = new double[]{meanW, meanW * meanW + varW};
        } else {
            double prec = Math.max(options.tol, 1e-14);
            FluFluResult fluResult = FluFluQueue.fluFluQueue(arrivalParams.Q, arrivalParams.R,
                    serviceParams.Q, serviceParams.R, true, 2, 2, prec);
            flMoms = fluResult.getFluidMoments();
            stMoms = fluResult.getSojournMoments();
            if (flMoms == null) throw new RuntimeException("FluFluQueue did not return fluid moments");
            if (stMoms == null) throw new RuntimeException("FluFluQueue did not return sojourn moments");
        }
        double meanQueueLength = flMoms[0];
        double meanResponseTime = stMoms[0];
        double throughput = (meanResponseTime > GlobalConstants.FineTol) ? meanQueueLength / meanResponseTime : lambda;
        double utilization = Math.min(1.0, lambda / mu);
        return new ClassResult(meanQueueLength, meanResponseTime, throughput, utilization);
    }

    private ProcessParams extractProcess(NetworkStruct sn, int stationIdx, int classIdx, double rate) {
        jline.lang.nodes.Station station = sn.stations.get(stationIdx);
        jline.lang.JobClass jobClass = sn.jobclasses.get(classIdx);
        java.util.Map<jline.lang.JobClass, MatrixCell> stationProc = sn.proc.get(station);
        MatrixCell proc = (stationProc != null) ? stationProc.get(jobClass) : null;
        if (proc == null || proc.size() == 0) {
            return new ProcessParams(Matrix.zeros(1, 1), Matrix.singleton(rate), true);
        }
        int nPhases = proc.get(0).getNumRows();
        if (nPhases == 1) {
            return new ProcessParams(Matrix.zeros(1, 1), Matrix.singleton(rate), true);
        }
        Matrix D0 = proc.get(0);
        Matrix D1 = proc.get(1);
        Matrix Q = D0.add(1.0, D1);
        Matrix R = Matrix.zeros(nPhases, nPhases);
        for (int i = 0; i < nPhases; i++) {
            double rowSum = 0.0;
            for (int j = 0; j < D1.getNumCols(); j++) rowSum += D1.get(i, j);
            R.set(i, i, rowSum);
        }
        return new ProcessParams(Q, R, false);
    }

    private static class TopologyInfo {
        boolean isValid = false;
        String errorMsg = "";
        int sourceNode = -1;
        int queueNode = -1;
        int sinkNode = -1;
        int sourceStation = -1;
        int queueStation = -1;
    }

    private static class ProcessParams {
        final Matrix Q;
        final Matrix R;
        final boolean isSimple;
        ProcessParams(Matrix Q, Matrix R, boolean isSimple) {
            this.Q = Q; this.R = R; this.isSimple = isSimple;
        }
    }

    private static class ClassResult {
        final double meanQueueLength;
        final double meanResponseTime;
        final double throughput;
        final double utilization;
        ClassResult(double meanQueueLength, double meanResponseTime, double throughput, double utilization) {
            this.meanQueueLength = meanQueueLength;
            this.meanResponseTime = meanResponseTime;
            this.throughput = throughput;
            this.utilization = utilization;
        }
    }
}

File: jar/src/main/kotlin/jline/solvers/fluid/analyzers/MFQAnalyzer.kt
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.solvers.fluid.analyzers

import jline.GlobalConstants
import jline.lang.NetworkStruct
import jline.lang.constant.JobClassType
import jline.lang.constant.NodeType
import jline.lib.butools.queues.fluFluQueue
import jline.solvers.SolverOptions
import jline.solvers.SolverResult
import jline.util.matrix.Matrix
import kotlin.math.min

/**
 * MFQ (Markovian Fluid Queue) analyzer for single-queue open systems.
 *
 * Uses BUTools FluFluQueue function to compute exact steady-state queue length
 * and sojourn time moments for single-queue networks with phase-type arrivals
 * and service distributions.
 *
 * Applicability:
 * - Single-queue open system: Source -> Queue -> Sink
 * - Single-server (c=1) or infinite-server (c=Inf)
 * - Exponential or phase-type distributions
 *
 * References:
 * Horvath G, Telek M, "Sojourn times in fluid queues with independent and
 * dependent input and output processes", PERFORMANCE EVALUATION 79: pp. 160-181, 2014.
 */
class MFQAnalyzer : FluidAnalyzer {
    private var xvecIt: Matrix? = null

    override fun analyze(sn: NetworkStruct, options: SolverOptions, result: SolverResult) {
        val M = sn.nstations
        val K = sn.nclasses

        // Initialize result matrices
        result.QN = Matrix.zeros(M, K)
        result.UN = Matrix.zeros(M, K)
        result.RN = Matrix.zeros(M, K)
        result.TN = Matrix.zeros(M, K)
        result.WN = Matrix.zeros(M, K)
        result.AN = Matrix.zeros(M, K)

        // Validate topology
        val topologyInfo = validateTopology(sn)
        if (!topologyInfo.isValid) {
            throw RuntimeException("MFQ requires single-queue topology: ${topologyInfo.errorMsg}")
        }

        val sourceStation = topologyInfo.sourceStation
        val queueStation = topologyInfo.queueStation

        // Process each open class
        for (k in 0 until K) {
            if (sn.njobs[0, k] == Double.POSITIVE_INFINITY) {
                // Open class
                val classResult = analyzeOpenClass(sn, k, sourceStation, queueStation, options)

                result.QN[queueStation, k] = classResult.meanQueueLength
                result.RN[queueStation, k] = classResult.meanResponseTime
                result.TN[queueStation, k] = classResult.throughput
                result.UN[queueStation, k] = classResult.utilization

                // Throughput at source equals queue throughput
                result.TN[sourceStation, k] = classResult.throughput
            }
        }

        // Create dummy state vector for compatibility
        xvecIt = Matrix.zeros(1, 1)
    }

    override fun getXVecIt(): Matrix? = xvecIt

    /**
     * Validates that the network has a single-queue topology suitable for MFQ.
     */
    private fun validateTopology(sn: NetworkStruct): TopologyInfo {
        val info = TopologyInfo()

        // Check for open classes
        var hasOpenClass = false
        for (k in 0 until sn.nclasses) {
            if (sn.njobs[0, k] == Double.POSITIVE_INFINITY) {
                hasOpenClass = true
                break
            }
        }
        if (!hasOpenClass) {
            info.errorMsg = "Not an open model - all classes are closed"
            return info
        }

        // Find Source, Queue, and Sink nodes
        var sourceNode = -1
        var queueNode = -1
        var sinkNode = -1
        var sourceCount = 0
        var queueCount = 0
        var sinkCount = 0

        for (i in 0 until sn.nnodes) {
            when (sn.nodetype[i]) {
                NodeType.Source -> {
                    sourceNode = i
                    sourceCount++
                }
                NodeType.Queue -> {
                    queueNode = i
                    queueCount++
                }
                NodeType.Sink -> {
                    sinkNode = i
                    sinkCount++
                }
                else -> {}
            }
        }

        // Validate counts
        when {
            sourceCount == 0 -> {
                info.errorMsg = "No source node found"
                return info
            }
            sourceCount > 1 -> {
                info.errorMsg = "Multiple source nodes found ($sourceCount)"
                return info
            }
            sinkCount == 0 -> {
                info.errorMsg = "No sink node found"
                return info
            }
            sinkCount > 1 -> {
                info.errorMsg = "Multiple sink nodes found ($sinkCount)"
                return info
            }
            queueCount == 0 -> {
                info.errorMsg = "No queue node found"
                return info
            }
            queueCount > 1 -> {
                info.errorMsg = "Multiple queue nodes found ($queueCount) - MFQ supports single queue only"
                return info
            }
        }

        // Get station indices
        val sourceStation = sn.nodeToStation[sourceNode].toInt()
        val queueStation = sn.nodeToStation[queueNode].toInt()

        // Check server count (single-server or infinite-server)
        val c = sn.nservers[queueStation, 0]
        if (c > 1 && c < Double.POSITIVE_INFINITY) {
            info.errorMsg = "Multi-server queue (c=$c) not supported - MFQ requires c=1 or c=Inf"
            return info
        }

        info.isValid = true
        info.sourceNode = sourceNode
        info.queueNode = queueNode
        info.sinkNode = sinkNode
        info.sourceStation = sourceStation
        info.queueStation = queueStation
        return info
    }

    /**
     * Analyzes an open class using FluFluQueue.
     */
    private fun analyzeOpenClass(
        sn: NetworkStruct,
        classIdx: Int,
        sourceStation: Int,
        queueStation: Int,
        options: SolverOptions
    ): ClassResult {
        // Extract arrival rate
        val lambda = sn.rates[sourceStation, classIdx]
        if (lambda.isNaN() || lambda <= 0) {
            throw RuntimeException("No valid arrival rate for class $classIdx at source")
        }

        // Extract service rate (mean)
        val mu = sn.rates[queueStation, classIdx]
        if (mu.isNaN() || mu <= 0) {
            throw RuntimeException("No valid service rate for class $classIdx at queue")
        }

        // Check stability
        val rho = lambda / mu
        if (rho >= 1.0) {
            // Unstable system
            return ClassResult(
                meanQueueLength = Double.POSITIVE_INFINITY,
                meanResponseTime = Double.POSITIVE_INFINITY,
                throughput = lambda,
                utilization = 1.0
            )
        }

        // Extract process matrices
        val arrivalParams = extractArrivalProcess(sn, sourceStation, classIdx, lambda)
        val serviceParams = extractServiceProcess(sn, queueStation, classIdx, mu)

        // Check if simple M/M/1 case
        val isSimpleExponential = arrivalParams.isSimple && serviceParams.isSimple

        val flMoms: DoubleArray
        val stMoms: DoubleArray

        if (isSimpleExponential) {
            // Use analytical M/M/1 formulas
            val meanL = rho / (1 - rho)
            val varL = rho / ((1 - rho) * (1 - rho))
            val meanW = 1.0 / (mu - lambda)
            val varW = 1.0 / ((mu - lambda) * (mu - lambda))

            flMoms = doubleArrayOf(meanL, meanL * meanL + varL)
            stMoms = doubleArrayOf(meanW, meanW * meanW + varW)
        } else {
            // Use FluFluQueue for general case
            val prec = maxOf(options.tol, 1e-14)
            val srv0stop = true  // Work-conserving behavior

            val fluResult = fluFluQueue(
                Qin = arrivalParams.Q,
                Rin = arrivalParams.R,
                Qout = serviceParams.Q,
                Rout = serviceParams.R,
                srv0stop = srv0stop,
                numFluidMoments = 2,
                numSojournMoments = 2,
                prec = prec
            )

            flMoms = fluResult.fluidMoments ?: throw RuntimeException("FluFluQueue did not return fluid moments")
            stMoms = fluResult.sojournMoments ?: throw RuntimeException("FluFluQueue did not return sojourn moments")
        }

        // Map moments to LINE metrics
        val meanQueueLength = flMoms[0]
        val meanResponseTime = stMoms[0]

        // Throughput from Little's Law: X = L / W
        val throughput = if (meanResponseTime > GlobalConstants.FineTol) {
            meanQueueLength / meanResponseTime
        } else {
            lambda
        }

        // Utilization: U = lambda * E[S]
        val utilization = min(1.0, lambda / mu)

        return ClassResult(
            meanQueueLength = meanQueueLength,
            meanResponseTime = meanResponseTime,
            throughput = throughput,
            utilization = utilization
        )
    }

    /**
     * Extracts arrival process parameters (Q, R matrices) from network structure.
     */
    private fun extractArrivalProcess(
        sn: NetworkStruct,
        stationIdx: Int,
        classIdx: Int,
        lambda: Double
    ): ProcessParams {
        // Get the station and job class objects
        val station = sn.stations[stationIdx]
        val jobClass = sn.jobclasses[classIdx]

        // Check if proc data is available
        val proc = sn.proc[station]?.get(jobClass)

        if (proc == null || proc.isEmpty()) {
            // Simple Poisson arrival (1-state)
            return ProcessParams(
                Q = Matrix.zeros(1, 1),
                R = Matrix.singleton(lambda),
                isSimple = true
            )
        }

        // Check number of phases
        val nPhases = proc[0].numRows
        if (nPhases == 1) {
            // Single state: Poisson arrival
            return ProcessParams(
                Q = Matrix.zeros(1, 1),
                R = Matrix.singleton(lambda),
                isSimple = true
            )
        }

        // Multi-phase: Extract D0, D1 and convert to (Q, R) format
        // D0 + D1 = Q (generator), sum of D1 row = R (fluid rate)
        val D0 = proc[0]
        val D1 = proc[1]

        val Q = D0.add(1.0, D1)
        val R = Matrix.zeros(nPhases, nPhases)
        for (i in 0 until nPhases) {
            var rowSum = 0.0
            for (j in 0 until D1.numCols) {
                rowSum += D1[i, j]
            }
            R[i, i] = rowSum
        }

        return ProcessParams(Q = Q, R = R, isSimple = false)
    }

    /**
     * Extracts service process parameters (Q, R matrices) from network structure.
     */
    private fun extractServiceProcess(
        sn: NetworkStruct,
        stationIdx: Int,
        classIdx: Int,
        mu: Double
    ): ProcessParams {
        // Get the station and job class objects
        val station = sn.stations[stationIdx]
        val jobClass = sn.jobclasses[classIdx]

        // Check if proc data is available
        val proc = sn.proc[station]?.get(jobClass)

        if (proc == null || proc.isEmpty()) {
            // Simple exponential service (1-phase)
            return ProcessParams(
                Q = Matrix.zeros(1, 1),
                R = Matrix.singleton(mu),
                isSimple = true
            )
        }

        // Check number of phases
        val nPhases = proc[0].numRows
        if (nPhases == 1) {
            // Single phase: exponential service
            return ProcessParams(
                Q = Matrix.zeros(1, 1),
                R = Matrix.singleton(mu),
                isSimple = true
            )
        }

        // Multi-phase: Extract D0, D1 and convert to (Q, R) format
        val D0 = proc[0]
        val D1 = proc[1]

        val Q = D0.add(1.0, D1)
        val R = Matrix.zeros(nPhases, nPhases)
        for (i in 0 until nPhases) {
            var rowSum = 0.0
            for (j in 0 until D1.numCols) {
                rowSum += D1[i, j]
            }
            R[i, i] = rowSum
        }

        return ProcessParams(Q = Q, R = R, isSimple = false)
    }

    /**
     * Internal data class for topology validation results.
     */
    private data class TopologyInfo(
        var isValid: Boolean = false,
        var errorMsg: String = "",
        var sourceNode: Int = -1,
        var queueNode: Int = -1,
        var sinkNode: Int = -1,
        var sourceStation: Int = -1,
        var queueStation: Int = -1
    )

    /**
     * Internal data class for process parameters.
     */
    private data class ProcessParams(
        val Q: Matrix,
        val R: Matrix,
        val isSimple: Boolean
    )

    /**
     * Internal data class for class analysis results.
     */
    private data class ClassResult(
        val meanQueueLength: Double,
        val meanResponseTime: Double,
        val throughput: Double,
        val utilization: Double
    )
}
